### PR TITLE
List/Retrieve instances via REST

### PR DESCRIPTION
**Pull request by abaiken**
Description: Enable users to access instances via REST. 
Issue: #133 
Demo: https://asciinema.org/a/NsXQil5gg2apKsMIhb5Jwo5Y4

### CODE REVIEW

File: cloudigrade/account/views.py [hunk 1]
```diff
@@ -1,10 +1,35 @@
...
-from account.models import Account, AwsAccount
+from account.models import (Account,
+                            AwsAccount,
+                            Instance)
+
+
+def convert_param_to_int(name, value):
+    """Check if a value is convertable to int.
+
+    :param name: The field name being validated
```
Comment (infinitewarp) [inline cloudigrade/account/views.py]: Use Google style docstrings.

We may move to the RST style later, but consistency wins for now.

File: cloudigrade/account/views.py [hunk 2]
```diff
@@ -25,7 +50,37 @@ def get_queryset(self):
...
+        if not user.is_superuser:
+            accounts = self.accounts.filter(user=user)
+            account_ids = [account.id for account in accounts]
+            return self.queryset.filter(account__in=account_ids)
+        user_id = self.request.query_params.get('user_id', None)
+        if user_id is not None:
+            user_id = convert_param_to_int('user_id', user_id)
+            accounts = self.accounts.filter(user__id=int(user_id))
+            account_ids = [account.id for account in accounts]
+            return self.queryset.filter(account__in=account_ids)
```
Comment (infinitewarp) [inline cloudigrade/account/views.py]: I'm pretty sure there's a cleaner way of doing this that doesn't require the "in list" like this, but I'll have to play with this and come back…

Comment (infinitewarp) [reply cloudigrade/account/views.py]: I played around in the shell with some local data, and this seems to work as desired:

```
>>> models.Instance.objects.filter(account__user__id=30)
<PolymorphicQuerySet [<AwsInstance i-0bee3c1c7c65b518c>, <AwsInstance i-05f78714782d7979b>]>
>>> models.Instance.objects.filter(account__user__id=31)
<PolymorphicQuerySet []>
```

So, I'd suggest simplifying `get_queryset` to look something like this:

```python
def get_queryset(self):
    """Get the queryset filtered to appropriate user."""
    user = self.request.user
    if not user.is_superuser:
        return self.queryset.filter(account__user__id=user.id)
    user_id = self.request.query_params.get('user_id', None)
    if user_id is not None:
        user_id = convert_param_to_int('user_id', user_id)
        return self.queryset.filter(account__user__id=user_id) 
    return self.queryset 
```

File: cloudigrade/account/views.py (hunk 1 above)
Comment (kholdaway) [reply cloudigrade/account/views.py]: Done.

File: cloudigrade/account/views.py (hunk 2 above)
Comment (kholdaway) [reply cloudigrade/account/views.py]: Great suggestion.  Done.

**Comment by kholdaway** [comment]
@infinitewarp Have another look.  Thanks for the feedback.

File: cloudigrade/account/tests/test_views.py [hunk 3]
```diff
@@ -221,3 +227,173 @@ def test_get_user1s_account_as_superuser_returns_ok(self):
...
+        self.factory = APIRequestFactory()
+
+    def assertResponseHasInstanceData(self, response, instance):
+        """Assert the response has data matching the instance object."""
+        self.assertEqual(
+            response.data['id'], instance.id
+        )
+        self.assertEqual(
+            response.data['account'],
+            'http://testserver/api/v1/account/' +
```
Comment (infinitewarp) [inline cloudigrade/account/tests/test_views.py]: This string concatenation could be collapsed to
```
f'http://testserver/api/v1/account/{instance.account.id}/'
```

**Comment by abaiken** [comment]
@infinitewarp Done.

File: cloudigrade/account/views.py [hunk 4]
```diff
@@ -1,10 +1,39 @@
...
 
 from account import serializers
 from account.exceptions import InvalidCloudProviderError
-from account.models import Account, AwsAccount
+from account.models import (Account,
+                            AwsAccount,
+                            Instance)
+
+
+def convert_param_to_int(name, value):
```
Comment (ilyavw) [inline cloudigrade/account/views.py]: This isn't a view, perhaps move it to `account.util`?

Comment (ilyavw) [reply cloudigrade/account/views.py]: If we plan to use it anywhere else, it's totally cool to move it to an appropriate place in `util.*`, maybe even a new package?

**Comment by kholdaway** [comment]
@infinitewarp @ilyavw back at you

File: cloudigrade/account/views.py [hunk 5]
```diff
@@ -1,10 +1,39 @@
...
+
+    """
+    if isinstance(value, int):
+        return value
+
+    try:
+        return int(value)
+    except ValueError:
+        error = {
+            name: [_('The id must be an integer.')]
```
[the remark this answers is not in the document]
Comment (kholdaway) [reply cloudigrade/account/views.py]: That syntax didn't work but I did something similar.

File: cloudigrade/account/views.py [hunk 6]
```diff
@@ -1,10 +1,39 @@
...
+        name (str): The field name being validated
+        value: The value to convert to int
+
+    Returns:
+        int: The int value
+    Raises:
+        ValidationError if value not convertable to an int
+
+    """
+    if isinstance(value, int):
```
[the remark this answers is not in the document]
Comment (kholdaway) [reply cloudigrade/account/views.py]: 👍

**Comment by kholdaway** [comment]
@infinitewarp Are the additional changes you need?

**Pull request merged**